### PR TITLE
[feat] Implement Cart Context and Functionality

### LINKED ISSUE #15: Implement cart context
The user can be allowed to add products directly to their carts, and thus it will reflect both in the cart count on navbar and update on the cart page as well, along with the price tally.

### PR DESCRIPTION
**Pull request by Hamdrive**
## Overview

This PR refers to issue #15

Successfully implemented context and overall logic/dynamics for cart page and also to both product card and cart product card

On the products page:
1.  User can add to cart, and on click button updates to directly go to cart
2. Count updates on navbar

On wishlist page:
1. If user adds product to wishlist, and from there adds to cart, button gets updated
2. Product gets updated on cart page, and simultaneously removed from wishlist page 

On Cart page:
1. User can increase the quantity of the product, and on decreasing less than 1, it removes the product
2. On increasing the quantity, the price tally also gets updated
3. User can move the product to wishlist, and both wishlist and cart gets updated

## Deployment

https://deploy-preview-19--fueledup-store.netlify.app/

## Preview

image
image

### CODE REVIEW

File: src/components/PriceSummary.jsx [hunk 1]
```diff
@@ -1,38 +1,64 @@
...
+        (Number(summary.initPrice) + Number(item.price)) * Number(item.quantity),
+      totalDisc:
+        (Number(summary.totalDisc) +
+          (Number(item.price) - Number(item.productDiscountPrice))) *
+        Number(item.quantity),
+      totalPrice:
+        (Number(summary.totalPrice) + Number(item.productDiscountPrice)) *
+        Number(item.quantity),
+        quantity: summary.quantity + item.quantity
+    }),
```
Comment (Varun-28) [inline src/components/PriceSummary.jsx:17]: > clear and understandable logic
> learned how we can do the discount feature
> good division of code among different files

File: src/utils/products/productsReducer.js [hunk 2]
```diff
@@ -11,12 +11,41 @@ export function productsReducer(state, action) {
...
+        return item;
+      });
+
+      return quantityZeroFlag
+        ? {
+            ...state,
+            cart: updatedCart.filter((item) => item._id !== action.payload._id),
+          }
+        : { ...state, cart: updatedCart };
     // case "MOVE_TO_WISHLIST":
```
Comment (saadahmadhere) [inline src/utils/products/productsReducer.js]: you can remove unwanted comments

File: src/components/PriceSummary.jsx [hunk 3]
```diff
@@ -2,18 +2,24 @@ import React from "react";
 
 export default function PriceSummary({ products }) {
   const tally = products.reduce(
-    (total, curr) => ({
-      ...total,
-      initPrice: Number(total.initPrice) + Number(curr.price),
+    (summary, item) => ({
+      ...summary,
+      initPrice:
+        (Number(summary.initPrice) + Number(item.price)) * Number(item.quantity),
```
Comment (saadahmadhere) [inline src/components/PriceSummary.jsx:8]: The discount thing was new to me. Learned a lot from your code.

**Review by hinanshisuthar — COMMENTED** [review]
Clean code practices followed. Nice implementation. Also, went through your preview, you may add a 'no items in cart/wishlist' when the wishlist/cart is empty. Other than that everything looks good to merge.

**Review by saadahmadhere — APPROVED** [review]
Very nice use of context, variables, and naming convention. Elegant.

File: src/utils/products/productsReducer.js (hunk 2 above)
Comment (Hamdrive) [reply src/utils/products/productsReducer.js]: Yes will cleanup the unused code

**Comment by Hamdrive** [comment]
Pull Request successfully reviewed by 3 reviewers, changes were suggested which have been implemented as well.
Additional changes suggested is outside the scope of this PR, and hence will be handled on a separate branch.
Thus this fixes issue #15 and PR will be merged and closed.

**Pull request merged**